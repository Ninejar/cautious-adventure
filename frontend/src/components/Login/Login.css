.content.login{
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    gap: 100px;
}

.content.login h1{
    font-size: 50pt;
}

.login_container{
    width: 400px;
    height: auto;
    margin: 0 auto;
    padding: 1em;
    background-color: rgb(244, 248, 247);
    display: flex;
    flex-direction: column;
    border-radius: 5px;
    gap: 30px;
    box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
}

.login_container input[type='email'],
.login_container input[type='password']{
    all: unset;
    background-color: rgb(46, 46, 46);
    padding: 1em;
    border-radius: 4px;
    border: solid 1px gray;
}

.login_container button{
    margin-top: 50px;
}

/* *{outline: 1px solid red} */