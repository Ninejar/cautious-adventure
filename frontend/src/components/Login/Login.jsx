import React, { useState } from 'react'
import axios from 'axios'
import {useNavigate} from 'react-router-dom'
import Loading from '../Loading'


const Login = () => {
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [loading, setLoading] = useState(false)
  const navigate = useNavigate()

  const handleLogin = async () => {
    setLoading(true);
    try {
      const response = await axios.post('http://localhost:1814/users/login', {
        email,
        password,
      });

      // Extract the token from the response
      const { token } = response.data;

      // Store the token in localStorage
      localStorage.setItem('auth-token', token);

      // Redirect to the home page or any other desired route
      navigate('/'); // Replace with your desired route
    } catch (error) {
      setLoading(false);
      console.error('Login error:', error);
      // Handle login error
    }
  };
  return (
    <div className='app'>
      <div className='content'>
      
      {loading ? <Loading /> : ''}

      <div className="create_journal">
        <input id="email" type="email" placeholder='email' value={email} onChange={(e) => setEmail(e.target.value)} />
        <input id="password" type="password" placeholder='password' value={password} onChange={(e) => setPassword(e.target.value)} />
         
        </div>

        <button onClick={handleLogin}>
          Login
        </button>
      </div>
      </div>
  )
}

export default Login