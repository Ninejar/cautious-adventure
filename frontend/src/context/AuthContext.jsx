import React, { createContext, useContext, useState, useEffect } from "react";
import { jwtDecode } from "jwt-decode";
import { useNavigate } from "react-router-dom";

// Create a context to store the user role
const AuthContext = createContext();

export const useAuth = () => useContext(AuthContext);

export const AuthProvider = ({ children }) => {
  // Store the user role in the state
  const [authState, setAuthState] = useState({
    userRole: "",
  });

  useEffect(() => {
    // Store token
    const token = localStorage.getItem("auth-token");
    if (token) {
      try {
        // Decode token to get user role
        const decoded = jwtDecode(token);
        // Update the user role in the state
        setAuthState({ userRole: decoded.role });
      } catch (error) {
        console.error("Error decoding token:", error);
        setAuthState({ userRole: "" });
      }
    } else {
      setAuthState({ userRole: "" });
    }
  }, []);

  const navigate = useNavigate();

  // Logout function
  const logout = () => {
    localStorage.removeItem("auth-token");
    setAuthState({ userRole: "" });
    navigate("/login");
  };

  // Login function
  const login = (token) => {
    // Store the response token in local storage
    localStorage.setItem("auth-token", token);
    const decoded = jwtDecode(token);
    setAuthState({ userRole: decoded.role });
  };

  return (
    // Provide the user role and the login/logout functions to the children
    <AuthContext.Provider
      value={{ userRole: authState.userRole, logout, login }}
    >
      {children}
    </AuthContext.Provider>
  );
};
