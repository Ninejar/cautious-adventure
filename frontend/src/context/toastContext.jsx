// ToastContext.js
import React, { createContext, useContext } from 'react';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

const ToastContext = createContext();

export const useToast = () => {
  return useContext(ToastContext);
};

export const ToastProvider = ({ children }) => {
  const showToast = (message) => {
    toast.success(message);
  };

  return (
    <ToastContext.Provider value={{ showToast }}>
      {children}
      <ToastContainer position='bottom-right' autoClose="1750"/>
    </ToastContext.Provider>
  );
};
