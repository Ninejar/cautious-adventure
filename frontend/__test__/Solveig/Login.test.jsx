import React from 'react';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { render, screen, fireEvent, waitFor, cleanup } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { AuthProvider } from "../../src/context/AuthContext";
import Login from '../../src/components/Login/Login';
import axios from 'axios';

// Mock axios and AuthContext for controlled testing environment
vi.mock('axios');
vi.mock('../../src/context/AuthContext', () => ({
    useAuth: () => ({ login: vi.fn() }),
    AuthProvider: ({ children }) => <div>{children}</div>
}));

describe('Login component - Verification of Expected Behavior - solveig', () => {
    beforeEach(() => {
        vi.resetAllMocks(); // Reset mocks to clean state before each test
        // dummy JWT for testing
        const dummyToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidGVhY2hlciIsImlhdCI6MTUxNjIzOTAyMn0.invalidsignature";
        axios.post.mockResolvedValue({ data: { token: dummyToken } });
    });

    afterEach(cleanup); // Clean up DOM after each test to prevent leakage

    it('renders only one login button - solveig', () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );
        const loginButtons = screen.getAllByTestId('login-button');
        expect(loginButtons.length).toBe(1);
    });

    it('sends correct data when login button is clicked with valid input - solveig', async () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );

        const emailInput = screen.getByPlaceholderText('Email');
        const passwordInput = screen.getByPlaceholderText('Password');
        const loginButton = screen.getByTestId('login-button');
        
        fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
        fireEvent.change(passwordInput, { target: { value: 'password123' } });
        fireEvent.click(loginButton);

        await waitFor(() => {
            expect(axios.post).toHaveBeenCalledWith('http://localhost:1814/users/login', {
                email: 'test@example.com',
                password: 'password123',
            });
        });
    });
    it('renders without crashing - solveig', () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );
         // Implicit test, successful render without errors
    });

    it('renders with custom props - solveig', () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login customProp="customValue" />
                </AuthProvider>
            </MemoryRouter>
        );
        // add props to check behaviors or effects of custom props
    });

    it('performs action on button click - solveig', () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );
        const button = screen.getByTestId('login-button');
        fireEvent.click(button); // Simulate button click and check resulting action
    });

    it('handles realistic usage cases - solveig', async () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );

        // Test with valid email and password
        fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: 'user@example.com' } });
        fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: 'password123' } });
        fireEvent.click(screen.getByTestId('login-button'));
        await waitFor(() => {
        });

        // Add more tests with different email and password combinations
        const emailAndPasswordCombinations = [
            { email: 'john@example.com', password: 'password123' },
            { email: 'jane@example.com', password: 'test@123' },
            { email: 'alice@example.com', password: 'al1c3P@ss!' },
            { email: 'bob@example.com', password: 'BoBPaSsW0rd!' },
        ];

        emailAndPasswordCombinations.forEach(({ email, password }) => {
            fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: email } });
            fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: password } });
            fireEvent.click(screen.getByTestId('login-button'));
        });
    });

    it('handles boundary cases - solveig', async () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );

        // Test with minimum and maximum password length
        fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: 'user@example.com' } });
        fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: 'a' } }); // Minimum length
        fireEvent.click(screen.getByTestId('login-button'));

        fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: '12345678901234567890123456789012345678901234567890' } }); // Maximum length
        fireEvent.click(screen.getByTestId('login-button'));
    });

    it('handles edge case - solveig', async () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );

        // Test with empty email and password
        fireEvent.click(screen.getByTestId('login-button'));

        // Test with extremely long email and password inputs
        fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: 'a'.repeat(256) } });
        fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: 'a'.repeat(256) } });
        fireEvent.click(screen.getByTestId('login-button'));
    });

    it('handles negative cases - solveig', async () => {
        render(
            <MemoryRouter>
                <AuthProvider>
                    <Login />
                </AuthProvider>
            </MemoryRouter>
        );

        // Test with invalid email formats
        const invalidEmails = ['user', 'user@', '@example.com'];
        invalidEmails.forEach(email => {
            fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: email } });
            fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: 'password123' } });
            fireEvent.click(screen.getByTestId('login-button'));
        });

        // Test with null or undefined email or password input
        fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: null } });
        fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: undefined } });
        fireEvent.click(screen.getByTestId('login-button'));

        // Test with SQL/NoSQL injection attempts
        fireEvent.change(screen.getByPlaceholderText('Email'), { target: { value: "'; DROP TABLE Users; --" } });
        fireEvent.change(screen.getByPlaceholderText('Password'), { target: { value: "'; DROP TABLE Users; --" } });
        fireEvent.click(screen.getByTestId('login-button'));
    });
});