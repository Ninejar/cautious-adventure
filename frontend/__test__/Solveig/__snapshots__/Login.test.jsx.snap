// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login component - Verification of Expected Behavior > matches snapshot 1`] = `
<div>
  <div>
    <div
      class="login_wrapper"
    >
      <div
        class="content login"
      >
        <h1
          data-testid="login-header"
        >
          Log in
        </h1>
        <div
          class="login_container"
        >
          <input
            id="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            id="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            data-testid="login-button"
          >
            Login
          </button>
        </div>
        <p>
          Don't have a user? Sign up 
          <a
            href="/signup"
          >
            here
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login component - Verification of Expected Behavior - solveig > matches snapshot - solveig 1`] = `
<div>
  <div>
    <div
      class="login_wrapper"
    >
      <div
        class="content login"
      >
        <h1
          data-testid="login-header"
        >
          Log in
        </h1>
        <div
          class="login_container"
        >
          <input
            id="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            id="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            data-testid="login-button"
          >
            Login
          </button>
        </div>
        <p>
          Don't have a user? Sign up 
          <a
            href="/signup"
          >
            here
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
